app/src/main/java/com/example/antitheft/MainActivity.java

```diff
@@ -0,0 +1,174 @@
+        SMSManager.getInstance().setReceivedListener(GPSCommandReceiver.class, context);
+
+        telephoneField = findViewById(R.id.telephoneNumber);
+        sendButton = findViewById(R.id.sendButton);
+        activity = this;
+
+        sendButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                boolean permissionsGranted = checkSelfPermission(Manifest.permission.SEND_SMS) == PERMISSION_GRANTED;
+                if (permissionsGranted) {
+                    String telephoneNumber = ((EditText) findViewById(R.id.telephoneNumber)).getText().toString();
+                    // Checking the peer.
+                    SMSPeer peer = new SMSPeer(telephoneNumber);
+                    if (peer == null)
+                        Toast.makeText(context, R.string.toast_non_valid_peer, Toast.LENGTH_SHORT).show();
+                        // If peer is not null, it sends the command.
+                    else sendCommand(new SMSMessage(peer, FULL_DEFAULT_LOCATE_COMMAND));
+                } else {
+                    // If SMS permissions are not granted, it will show a toast, because
+                    // the app can't send commands without that permission.
+                    Toast.makeText(context, R.string.toast_app_needs_SMS_permissions, Toast.LENGTH_SHORT).show();
+                    requestPermissions(new String[]{Manifest.permission.SEND_SMS}, REQUEST_CODE);
+                }
+            }
+        });
+        // Request permissions.
+        requestPermissions(PERMISSIONS, REQUEST_CODE);
+
+        // Immediately prepare the last known location.
+        client = LocationServices.getFusedLocationProviderClient(this);
+        // Callback to retrieve the last known location.
+        client.getLastLocation().addOnSuccessListener(this, new OnSuccessListener<Location>() {
+            @Override
+            public void onSuccess(Location location) {
+                locationParser.setLocation(location);
+            }
+        });
+
+    }
+
+    /**
+     * This method sends the command, calling the method in {@link GPSCommandHandler} class, if
+     * command is not null.
+     *
+     * @param smsMessage The message to send.
+     * @see com.eis.smslibrary.SMSMessage
+     * @see GPSCommandHandler
+     * @see GPSCommandHandler#sendMessage(SMSMessage)
+     */
+    private void sendCommand(@Nullable SMSMessage smsMessage) {
+        if (smsMessage != null)
+            new GPSCommandHandler().sendMessage(smsMessage);
+        else
+            Toast.makeText(this, R.string.sendCommand_fail, Toast.LENGTH_SHORT).show();
+    }
+
+    /**
+     * This method retrieves the last known {@link Location}, parsed as String.
+     *
+     * @return The last known location.
+     * @see Location
+     * @see LocationParser
+     * @see LocationParser#toString()
+     */
+    public static String getCurrentLocation() {
+        String locationTag = "LOCATION";
```

Comment:
This could be a final variable.

Reply:
Oh shoot, I forgot to delete it!
Thank you for spotting that.

Actually, it wasn't supposed to be final, because it was supposed to be used, for debug reason, as a tag in `Log.d(String tag, String message)` inside `getCurrentLocation()` only.
So, since there were 2 Logs, in order to give them a visible category, without hardwiring, I decided this way.